packages/contracts/src/contractManager.ts

```diff
@@ -0,0 +1,122 @@
+import type { AssetId, ChainId } from '@shapeshiftoss/caip'
+import { fromAssetId, toAssetId } from '@shapeshiftoss/caip'
+import type { PartialRecord } from '@shapeshiftoss/types'
+import { KnownChainIds } from '@shapeshiftoss/types'
+import type { Token } from '@uniswap/sdk'
+import { Fetcher } from '@uniswap/sdk'
+import assert from 'assert'
+import memoize from 'lodash/memoize'
+import type { Address } from 'viem'
+import { getAddress, getContract } from 'viem'
+
+import { CONTRACT_ADDRESS_TO_ABI, CONTRACT_TYPE_TO_ABI } from './constants'
+import { getEthersV5Provider } from './ethersProviderSingleton'
+import type {
+  KnownContract,
+  KnownContractAbiByAddress,
+  KnownContractAddress,
+  KnownContractByAddress,
+  KnownContractByType,
+} from './types'
+import { ContractType } from './types'
+import { viemClientByChainId, viemEthMainnetClient } from './viemClient'
+
+const definedContracts: PartialRecord<Address, KnownContract> = {}
+
+export const getOrCreateContractByAddress: <A extends KnownContractAddress>(
+  address: A,
+) => KnownContractByAddress<A> = <A extends KnownContractAddress>(
+  address: A,
+): KnownContractByAddress<A> => {
```

Comment:
not sure why the more verbose function declaration versus the original implementation version?

also nitpick - leverage more standard naming convention of generic types with `T` over `A`

Reply:
playing with it a little bit looks like some kind of type hell going on with these functions, so that probably explains the difference...

Reply:
Exactly, the typings weren't complete here and its took a fair bit of fandangling to get them to line up